core/src/toga/images.py

```diff
@@ -1,52 +1,104 @@
-            raise ValueError("Either path or data must be set.")
-        if path is not None and data is not None:
-            raise ValueError("Only either path or data can be set.")
-
+        # At first we will create a list with these three variable, then count how many None is in that list.
+        # If the number of None is 1 -> raises ValueError. One and only One of the three variables must be set but here two of them are set
+        # If the number of None is 3 -> raises ValueError. One and only One of the three variables must be set but here none of them are set 
+        # If the number of None is 2 -> Ok. Check the validity of the value of that non-None variable
+        none_count = [path, data, pil_image].count(None)
+        if none_count != 2:
+            raise ValueError("One and Only one of the three args (path, data, pil_image) must be set.")
+        # checking validity of the arg(one of the three)
         if path is not None:
             if isinstance(path, Path):
                 self.path = path
             else:
                 self.path = Path(path)
             self.data = None
-        else:
-            self.path = None
+            self.pil_image = None
+        if data is not None:
             self.data = data
+            self.path = None
+            self.pil_image = None
+        if pil_image is not None:
+            try:
+                from PIL import Image as _PIL_Image_Module_
+            except ImportError as e:
+                print("Pillow is required as pil_image is set to a PIL.Image type")
+                raise ImportError(e)
+            #type checking
+            if not _PIL_Image_Module_.isImageType(pil_image):
+                raise TypeError("pil_image is not a PIL.Image type.")
+            self.pil_image = pil_image
+            self.data = None
+            self.path = None
+
 
         self.factory = get_platform_factory()
         if self.data is not None:
             self._impl = self.factory.Image(interface=self, data=self.data)
-        else:
+        if self.path is not None:
             self.path = toga.App.app.paths.app / self.path
             if not self.path.is_file():
                 raise FileNotFoundError(f"Image file {self.path} does not exist")
             self._impl = self.factory.Image(interface=self, path=self.path)
+        if self.pil_image is not None:
+            #What we will do: save the pil_image into a temporary file then load raw byte data
+            #from that file, then use the bytes to create self._impl, finally delete the temp file
+
+            # creating an unique name for the temporary file
+            unique_time = str(time.time())
+            temp_file = f"_PIL_temp_img_file_{unique_time}_"
+            temp_file_path = toga.App.app.paths.app
+            while temp_file in os.listdir(temp_file_path):
+                temp_file+="1_"
+            complete_temp_path = os.path.join(temp_file_path, temp_file)
```

Comment:
If you're writing a file to disk, you're doing it wrong. There is no need to use a file on disk to load an image.